activerecord/test/cases/tasks/database_tasks_test.rb

```diff
@@ -467,6 +467,17 @@ class DatabaseTasksStructureLoadTest < ActiveRecord::TestCase
   end
 
   class DatabaseTasksCheckSchemaFileTest < ActiveRecord::TestCase
+    def setup
+      unless Rails.respond_to?(:root)
+        @defined_root = true
+        def Rails.root; :defined_root; end
```

Comment:
iirc, Active Support has a `redefine_possible_method` method we might be able to use here.